File: frontend/src/app/core/services/admin-marketplace.service.ts
import { Injectable, inject } from '@angular/core';
import { ApiService } from './api.service';
import { pageParams } from '../../shared/utils/page-params.util';
import {
  Listing,
  Review,
  ListingDetail,
  MarketplaceKpis,
  ListingFilters,
  ReviewFilters,
  Page,
} from '../../models/admin-marketplace.model';

/**
 * Platform-staff API client for Marketplace Moderation: listing + review lists (filterable +
 * paginated), KPI totals, listing detail, and the moderation mutations (take down / restore /
 * feature, hide / unhide / delete). All gated by MARKETPLACE_MODERATE + audit-logged on the backend.
 */
@Injectable({ providedIn: 'root' })
export class AdminMarketplaceService {
  private readonly api = inject(ApiService);
  private readonly base = '/admin/marketplace';

  listings(filters: ListingFilters = {}, page = 0, size = 10) {
    const params = pageParams(filters, page, size);
    return this.api.get<Page<Listing>>(`${this.base}/listings`, { params });
  }

  reviews(filters: ReviewFilters = {}, page = 0, size = 10) {
    const params = pageParams(filters, page, size);
    return this.api.get<Page<Review>>(`${this.base}/reviews`, { params });
  }

  kpis() {
    return this.api.get<MarketplaceKpis>(`${this.base}/kpis`);
  }

  getListing(id: string) {
    return this.api.get<ListingDetail>(`${this.base}/listings/${id}`);
  }

  takeDown(id: string, reason: string | null) {
    return this.api.post<Listing>(`${this.base}/listings/${id}/takedown`, { reason });
  }
  restore(id: string) { return this.api.post<Listing>(`${this.base}/listings/${id}/restore`, {}); }
  feature(id: string) { return this.api.post<Listing>(`${this.base}/listings/${id}/feature`, {}); }
  unfeature(id: string) { return this.api.post<Listing>(`${this.base}/listings/${id}/unfeature`, {}); }

  hideReview(id: string) { return this.api.post<Review>(`${this.base}/reviews/${id}/hide`, {}); }
  unhideReview(id: string) { return this.api.post<Review>(`${this.base}/reviews/${id}/unhide`, {}); }
  deleteReview(id: string, reason: string | null) {
    const q = reason ? `?reason=${encodeURIComponent(reason)}` : '';
    return this.api.delete<void>(`${this.base}/reviews/${id}${q}`);
  }
}

File: backend/src/main/java/com/postwerk/service/AdminMarketplaceService.java
package com.postwerk.service;

import com.postwerk.dto.admin.AdminListingDetailResponse;
import com.postwerk.dto.admin.AdminListingResponse;
import com.postwerk.dto.admin.AdminReviewResponse;
import com.postwerk.dto.admin.MarketplaceKpisResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

/**
 * Platform-staff Marketplace Moderation: cross-tenant listing + review moderation (take down / restore
 * / feature, hide / unhide / delete), gated by {@code MARKETPLACE_MODERATE} (enforced at the controller).
 *
 * @since 1.0
 */
public interface AdminMarketplaceService {

    Page<AdminListingResponse> listListings(String search, String status, String kind, String pricing,
                                            String sort, Pageable pageable);

    Page<AdminReviewResponse> listReviews(String search, Integer rating, String status, String sort, Pageable pageable);

    MarketplaceKpisResponse kpis();

    AdminListingDetailResponse getListing(UUID listingId);

    AdminListingResponse takeDown(UUID listingId, String reason, UUID actorUserId, String ip);

    AdminListingResponse restore(UUID listingId, UUID actorUserId, String ip);

    AdminListingResponse setFeatured(UUID listingId, boolean featured, UUID actorUserId, String ip);

    AdminReviewResponse setReviewHidden(UUID reviewId, boolean hidden, UUID actorUserId, String ip);

    void deleteReview(UUID reviewId, String reason, UUID actorUserId, String ip);
}
